pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,337 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package appconfigreporter
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+
+	"github.com/pipe-cd/pipe/pkg/app/api/service/pipedservice"
+	"github.com/pipe-cd/pipe/pkg/config"
+	"github.com/pipe-cd/pipe/pkg/git"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+type apiClient interface {
+	UpdateApplicationConfigurations(ctx context.Context, in *pipedservice.UpdateApplicationConfigurationsRequest, opts ...grpc.CallOption) (*pipedservice.UpdateApplicationConfigurationsResponse, error)
+	ReportUnregisteredApplicationConfigurations(ctx context.Context, in *pipedservice.ReportUnregisteredApplicationConfigurationsRequest, opts ...grpc.CallOption) (*pipedservice.ReportUnregisteredApplicationConfigurationsResponse, error)
+}
+
+type gitClient interface {
+	Clone(ctx context.Context, repoID, remote, branch, destination string) (git.Repo, error)
+}
+
+type applicationLister interface {
+	List() []*model.Application
+}
+
+type environmentGetter interface {
+	GetByName(ctx context.Context, name string) (*model.Environment, error)
+}
+
+type Reporter struct {
+	apiClient         apiClient
+	gitClient         gitClient
+	applicationLister applicationLister
+	envGetter         environmentGetter
+	config            *config.PipedSpec
+	gitRepos          map[string]git.Repo
+	gracePeriod       time.Duration
+	// Cache for the last scanned commit for each repository.
+	// Not goroutine safe.
+	lastScannedCommits map[string]string
+	fsys               fs.FS
+	logger             *zap.Logger
+}
+
+func NewReporter(
+	apiClient apiClient,
+	gitClient gitClient,
+	appLister applicationLister,
+	envGetter environmentGetter,
+	cfg *config.PipedSpec,
+	gracePeriod time.Duration,
+	logger *zap.Logger,
+) *Reporter {
+	return &Reporter{
+		apiClient:          apiClient,
+		gitClient:          gitClient,
+		applicationLister:  appLister,
+		envGetter:          envGetter,
+		config:             cfg,
+		gracePeriod:        gracePeriod,
+		lastScannedCommits: make(map[string]string),
+		// FIXME: Find another way to specify the root dir for a file system
+		fsys:   os.DirFS("/"),
```

Comment:
Initializing this with `/` is not a best practice so I think we should avoid this.

How about defining our own FS like

```
type fileSystem struct {}

func (s *fileSystem) Open(name string) (File, error) {
   return os.Open(name)
}

```

Reply:
Love it! Let me apply that simpler one.